### PR TITLE
Ability to get scores for questions with multiple rows on SurveyMonkey agent

### PR DESCRIPTION
**Pull request by monsha**
To get score when answers are multi rows, the agent looks for every choice_id in survey details and extract the value, then the average of all choices is calculated. Similar for comments.

https://trello.com/c/G68h8Lfu/475-add-ability-to-get-scores-for-questions-with-multiple-rows-on-surveymonkey-agent

### Event payload
Before:
```
{
  "score": 2,
  "comment": "more dresses for telephone numbers say more in contact with your customer",
  "id": "5431616232",
  "survey_id": "60926882",
  "date_created": "2017-09-23T14:56:29+00:00",
  "collector_id": "75862972",
  "custom_variables": {
  },
  "analyze_url": "https://www.surveymonkey.com/analyze/browse/rnYKfhMmvoohXEONXlb_2FTXSdkotX6xc5BfUDlX97ohg_3D?respondent_id=5431616232",
  "language": "en"
}
```
After:
```
{
  "score": 2,
  "comment": "one comment\nTwo comment",
  "id": "5431616232",
  "survey_id": "60926882",
  "date_created": "2017-09-23T14:56:29+00:00",
  "collector_id": "75862972",
  "custom_variables": {
  },
  "analyze_url": "https://www.surveymonkey.com/analyze/browse/rnYKfhMmvoohXEONXlb_2FTXSdkotX6xc5BfUDlX97ohg_3D?respondent_id=5431616232",
  "language": "en"
}
```

### CODE REVIEW

File: app/services/survey_monkey_parser.rb [hunk 1]
```diff
@@ -112,12 +117,15 @@ def comment_question
...
 
-    def comment_answer_given
-      comment_question['answers'].first
+    def comment_answers_given
+      comment_question['answers']
     end
 
     def parsed_comment_answer
-      comment_answer_given['text']
+      comments = comment_answers_given.map do |answ|
```
Comment (OpakAlex) [inline app/services/survey_monkey_parser.rb]:
```
comment_answers_given.map { |answ|  answ['text'] }.join("\n")
```

File: app/services/survey_monkey_parser.rb [hunk 2]
```diff
@@ -96,12 +96,17 @@ def score_options
...
-    def score_answer_given
-      score_question['answers'].first['choice_id']
+    def score_answers_given
+      score_question['answers']
     end
 
-    def parsed_score_answer
-      score_options.find { |c| c['id'] == score_answer_given }['text'].gsub(/[^0-9]/, '').to_i
+    def parsed_score_answer(choice_id)
+      score_options.find { |c| c['id'] == choice_id }['text'].gsub(/[^0-9]/, '').to_i
```
Comment (OpakAlex) [inline app/services/survey_monkey_parser.rb]: maybe we need to check if find found something?

**Review by OpakAlex — DISMISSED** [review]

File: app/services/survey_monkey_parser.rb [hunk 3]
```diff
@@ -101,11 +101,17 @@ def score_answers_given
     end
 
     def parsed_score_answer(choice_id)
-      score_options.find { |c| c['id'] == choice_id }['text'].gsub(/[^0-9]/, '').to_i
+      choice = score_options.find { |c| c['id'] == choice_id }
+
+      if choice['weight'].present?
```
Comment (monsha) [inline app/services/survey_monkey_parser.rb]: @duboff I think weight is always present and there are some surveys where we won't get the correct value, such as the survey `22873544` (https://api.surveymonkey.net/v3/surveys/122873544/details), weight field will return in this case `-100`

**Comment by duboff** [comment]
No, not everywhere, but you are right. It can be completely misleading. This is maddening

**Comment by duboff** [comment]
Still WIP

File: app/services/survey_monkey_parser.rb [hunk 4]
```diff
@@ -70,13 +74,42 @@ def parse
...
+      questions.each_with_object({}) do |question, hsh|
+        details = question.dig('details')
+        heading = details['headings']&.first['heading']
+
+        hsh[question['id']] = {
+          id: question['id'],
+          family: details['family'],
+          subtype: details['subtype'],
+          question: heading,
+          answers: question['answers']&.each_with_object({}) do |a, hsh|
```
Comment (OpakAlex) [inline app/services/survey_monkey_parser.rb]: maybe we can separate into method get answers?

File: app/services/survey_monkey_parser.rb [hunk 5]
```diff
@@ -96,12 +129,23 @@ def score_options
...
+      else
+        choice['text'].gsub(/[^0-9]/, '').to_i
+      end
     end
 
-    def parsed_score_answer
-      score_options.find { |c| c['id'] == score_answer_given }['text'].gsub(/[^0-9]/, '').to_i
+    def average_score
+      total = score_answers_given.reduce(0) { |sum, answer| sum + parsed_score_answer(answer["choice_id"]) }
+      total / (score_answers_given.size.nonzero? || 1)
```
Comment (OpakAlex) [inline app/services/survey_monkey_parser.rb]: `score_answers_given.size.nonzero?` return true / false? If yes this code is incorrect

**Review by OpakAlex — CHANGES_REQUESTED** [review]

**Comment by duboff** [comment]
@OpakAlex not yet ready for review :) but good point

File: app/services/survey_monkey_parser.rb (hunk 5 above)
Comment (duboff) [reply app/services/survey_monkey_parser.rb]: it returns `nil` / `true` I think

Comment (monsha) [reply app/services/survey_monkey_parser.rb]: It returns `nil` or `score_answers_given.size`

File: app/models/agents/survey_monkey_agent.rb [hunk 6]
```diff
@@ -169,6 +166,17 @@ def per_page
       interpolated['per_page']
     end
 
+    def build_survey(survey_id)
+      survey = fetch_survey_details(survey_id)
+      survey['responses'] = fetch_survey_responses(survey_id)
+      if boolify(interpolated['guess_mode']) == false
```
Comment (OpakAlex) [inline app/models/agents/survey_monkey_agent.rb:172]: maybe just unless?

**Review by OpakAlex — APPROVED** [review]
Looks good, just one small comment about logic style

File: app/models/agents/survey_monkey_agent.rb (hunk 6 above)
Comment (monsha) [reply app/models/agents/survey_monkey_agent.rb:172]: @OpakAlex boolify can return false/nil, in this case needs to be explicitly false

**Review by duboff — APPROVED** [review]

**Pull request merged**